src/QirRuntime/lib/QIR/arrays.cpp

```diff
@@ -14,35 +14,41 @@
 
 #include "CoreTypes.hpp"
 
-long QirArray::AddRef()
+int QirArray::AddRef()
 {
-    long rc = ++this->refCount;
-    assert(rc != 1); // not allowed to resurrect!
-    return rc;
+    assert(this->refCount != 0 && "Cannot resurrect released array!");
+    return ++this->refCount;
 }
 
 // NB: Release doesn't trigger destruction of the QirArray itself to allow for it
 // being used both on the stack and on the heap. The creator of the array
 // should delete it, if allocated from the heap.
-long QirArray::Release()
+int QirArray::Release()
 {
-    const long rc = --this->refCount;
+    assert(this->refCount != 0 && "Cannot release already released array!");
+    const int rc = --this->refCount;
     if (rc == 0)
     {
-        if (this->ownsQubits)
-        {
-            QUBIT** qubits = reinterpret_cast<QUBIT**>(this->buffer);
-            for (long i = 0; i < this->count; i++)
-            {
-                quantum__rt__qubit_release(qubits[i]);
-            }
-        }
         delete[] this->buffer;
         this->buffer = nullptr;
     }
     return rc;
 }
 
+void QirArray::AddUser()

```

Comment:
Is there a place where it is documented what is the meaning of a user in the context of QIR arrays, and how it is different from references?

Reply:
Yes. The QIR specification: https://github.com/microsoft/qsharp-language/tree/main/Specifications/QIR (@bettinaheim has a pending PR there that explains user counts).